rllib/core/rl_module/marl_module.py

```diff
@@ -0,0 +1,295 @@
+
+        Args:
+            state_dict: The state dict to set.
+        """
+        for module_id, module in self._rl_modules.items():
+            module.set_state(state_dict[module_id])
+
+    @override(RLModule)
+    def as_multi_agent(self) -> "MultiAgentRLModule":
+        """Returns a multi-agent wrapper around this module.
+
+        This method is overridden to avoid double wrapping.
+
+        Returns:
+            The instance itself.
+        """
+        return self
+
+    def add_module(
+        self,
+        module_id: ModuleID,
+        module: RLModule,
+        *,
+        dist_config=None,
+        override: bool = False,
+    ) -> None:
+        """Adds a module at run time to the multi-agent module.
+
+        Args:
+            module_id: The module ID to add. If the module ID already exists and
+                override is False, an error is raised. If override is True, the module
+                is replaced.
+            module: The module to add.
+            dist_config: The distributed configuration to use if module needs to be
+                distributed.
+            override: Whether to override the module if it already exists.
+
+        Raises:
+            ValueError: If the module ID already exists and override is False.
+        """
+        if module_id in self._rl_modules and not override:
+            raise ValueError(
+                f"Module ID {module_id} already exists. If your intention is to "
+                "override, set override=True."
+            )
+        if self.is_distributed():
+            module.make_distributed(dist_config)
+        self._rl_modules[module_id] = module
+
+    def remove_module(
+        self, module_id: ModuleID, *, raise_err_if_not_found: bool = True
+    ) -> None:
+        """Removes a module at run time from the multi-agent module.
+
+        Args:
+            module_id: The module ID to remove.
+            raise_err_if_not_found: Whether to raise an error if the module ID is not
+                found.
+        Raises:
+            ValueError: If the module ID does not exist and raise_err_if_not_found is
+                True.
+        """
+        if raise_err_if_not_found:
+            self._check_module_exists(module_id)
+        del self._rl_modules[module_id]
+
+    def __getitem__(self, module_id: ModuleID) -> RLModule:
+        """Returns the module with the given module ID.
+
+        Args:
+            module_id: The module ID to get.
+
+        Returns:
+            The module with the given module ID.
+        """
+        self._check_module_exists(module_id)
+        return self._rl_modules[module_id]
+
+    @OverrideToImplementCustomLogic
+    def _make_modules(self) -> Mapping[ModuleID, RLModule]:
+        modules_info_dict = self.config["modules"]
+        modules = {}
+        for module_id, module_info in modules_info_dict.items():
+
+            if isinstance(module_info, RLModule):
+                modules[module_id] = module_info
```

Comment:
wait, how do people specify an initialized RLModule in the config?

Reply:
They have three options specified in the docstring of the class. 1) instance of RLModule 2) dict containing class and config 3) a tuple of class and configs. 2 gives more flexibility and being explicit, 3 will give better user-experience. I am keeping both for now :)

Reply:
ps: I removed the tuple way of doing this. I feel like being able to specify the class and configs using a dict is enough

Reply:
👌